base/iterators.jl

```diff
@@ -957,4 +958,99 @@ function next(itr::PartitionIterator, state)
     return resize!(v, i), state
 end
 
+"""
+    Stateful(itr)
+
+There are several different ways to think about this iterator wrapper:
+    1. It provides a mutable wrapper around an iterator and
+       its iteration state.
+    2. It turns an iterator-like abstraction into a Channel-like
+       abstraction.
+    3. It's an iterator that mutates to become its own rest iterator
+       whenever an item is produced.
+
+`Stateful` provides the regular iterator interface. Like other mutable iterators
+(e.g. `Channel`), if iteration is stopped early (e.g. by a `break` in a `for` loop),
+iteration can be resumed from the same spot by continuing to iterate over the
+same iterator object (in contrast, an immutable iterator would restart from the
+beginning).
+
+# Example:
+```jldoctest
+julia> a = Iterators.Stateful("abcdef");
+
+julia> isempty(a)
+false
+
+julia> take!(a)
+'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
+
+julia> collect(Iterators.take(a, 3))
+3-element Array{Char,1}:
+ 'b'
+ 'c'
+ 'd'
+
+julia> collect(a)
+2-element Array{Char,1}:
+ 'e'
+ 'f'
+```
+
+```jldoctest
+julia> a = Iterators.Stateful([1,1,1,2,3,4]);
+
+julia> for x in a; x == 1 || break; end
+
+julia> Base.peek(a)
+3
```

Comment:
why not overloads `first`, like `zip` does?

Reply:
Fair point, that would be consistent, though the generic definition of `first` iterates, so people might expect it to consume an item (that doesn't really bother me too much though, since `take!` is the consume-one-item function). `first` is defined to error on an empty collection though, I find returning `nothing` a more useful behavior, but I guess `peek` on IO doesn't really do that either. Maybe we should have a `first` equivalent that returns nothing or another sentinel when the collection is empty? @nalimilan ?

Reply:
> Fair point, that would be consistent, though the generic definition of `first` iterates, so people might expect it to consume an item.

I don't think `first` will consume item given that it's not `first!`.
> `first` is defined to error on an empty collection though

We already have something breaks that rule.
```julia
julia> first(1:0)                                                                                                                                 
1
```

Reply:
Changing the state of stateful iterators and doing I/O has never qualified as mutation in the sense that `!` at the end of a name indicates. Otherwise all of the I/O functions would end in `!` yet they do not. We need to decide if `first` qualifies as iteration or not. Having `peek` as a non-iterating form of `take` seems reasonable to me, but then so would `first`.

Reply:
I'd also expect `first` to consume elements, so I think I prefer adding `peek`. We could generalize it so that it returns `nothing` for empty collections as @Keno suggests. Also, it sounds like one function is redundant in the `first`/`peek`/`take!` triad, and I'd tend to drop `take!` in favor of `first`.

Regarding the empty ranges issue at https://github.com/JuliaLang/julia/pull/25385, I guess we could make `first` return `nothing` in general for empty collections, but I'm a bit concerned about the performance implications.

Reply:
We could use `popfirst!` instead of `take!` for the mutating operation. I still like `peek` rather than `first`, because the latter feels like a bit of a pun.